dali/base/dasds.cpp

```diff
@@ -2095,6 +2101,146 @@ ISDSManagerServer &querySDSServer()
 
 /////////////////
 
+void CExternalFile::remove(const char *name)
+{
+    StringBuffer filename(name);
+    filename.append(ext);
+    manager.extCache.remove(filename);
+    manager.deltaWriter.removeExt(filename.detach());
+}
+
+void CBinaryFileExternal::readValue(const char *name, MemoryBuffer &mb)
+{
+    StringBuffer extName;
+    getName(extName, name);
+    if (manager.extCache.lookup(extName, mb))
+        return;
+
+    StringBuffer filename;
+    getFilename(filename, name);
+    Owned<IFile> iFile = createIFile(filename.str());
+    size32_t sz = (size32_t)iFile->size();
+    if ((unsigned)-1 == sz)
+    {
+        StringBuffer s("Missing external file ");
+        Owned<IException> e = MakeSDSException(SDSExcpt_MissingExternalFile, "%s", filename.str());
+        LOG(MCoperatorWarning, unknownJob, e, s.str());
+        StringBuffer str("EXTERNAL BINARY FILE: \"");
+        str.append(filename.str()).append("\" MISSING");
+        CPTValue v(str.length()+1, str.str(), false);
+        v.serialize(mb);
+    }
+    else
+    {
+        Owned<IFileIO> fileIO = iFile->open(IFOread);
+        verifyex(sz == ::read(fileIO, 0, sz, mb));
+        manager.extCache.add(extName, sz, mb.toByteArray());
+    }
+}
+
+void CBinaryFileExternal::read(const char *name, IPropertyTree &owner, MemoryBuffer &mb, bool withValue)
+{
+    const char *_name = owner.queryName();
+    if (!_name) _name = "";
+    mb.append(_name);
+
+    size32_t flagsPos = mb.length();
+    byte flags = ((PTree &)owner).queryFlags();
+    mb.append(flags);
+    serializeVisibleAttributes(owner, mb);
+    StringBuffer extName;
+    getName(extName, name);
+    if (manager.extCache.lookup(extName, mb))
```

Comment:
Should withValue be passed/tested before this?  Is it something that occurs often?

Reply:
>Should withValue be passed/tested before this?

it does look that way, I need to double check.

> Is it something that occurs often?

is hitting the cache something that happens often?
Yes, in testing, the same external is often re-read in quick succession.